File: 46-permutations/permutations.java
class Solution {

    public void help(Map<Integer, Integer> mp, List<List<Integer>> lst, List<Integer> curr, int[] nums) {
        if (curr.size() == nums.length) {
            lst.add(new ArrayList<>(curr));
            return;
        }
        for (int i : nums) {
            if (!mp.containsKey(i)) {
                curr.add(i);
                mp.put(i, 0);

                help(mp, lst, curr, nums);

                curr.remove(curr.size() - 1);
                mp.remove(i);
            }
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> lst = new ArrayList<>();
        Map<Integer, Integer> mp = new HashMap<>();

        help(mp, lst, new ArrayList<>(), nums);

        return lst;
    }
}

File: 46-permutations/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def helper(idx):
            if idx == len(nums):
                res.append(nums[:])
                return

            for i in range(idx, len(nums)):
                nums[i], nums[idx] = nums[idx], nums[i]
                helper(idx + 1)
                nums[i], nums[idx] = nums[idx], nums[i]

        res = []
        helper(0)
        return res
